include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -633,14 +633,17 @@ float irr_glsl_blinn_phong_cos_eval_DG(in irr_glsl_LightSample s, in irr_glsl_An
 // sample and microfacet are inout because of handleTwosided() only
 void instr_eval_and_pdf_execute(in instr_t instr, in MC_precomputed_t precomp, in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache _microfacet)
 {
-	uint op = instr_getOpcode(instr);
+	const uint op = instr_getOpcode(instr);
 
```

Comment:
you forgot to clamp all the `dot` variables before using them...

Also a quick check for the **after clamp value** of `NdotL` coudl save you a lot of computation here (and data fetches)

Reply:
> Also a quick check for the **after clamp value** of `NdotL` coudl save you a lot of computation here (and data fetches)

Applicable to BRDF leafs only.

Only stuff that has nothing to do with the PDF computation (so all data except for roughness) and computation of Fresnel and G2.

Reply:
ok looks good, checked it all